### PR TITLE
mkcloud: pass proxy variables into admin node

### PR DESCRIPTION
**Pull request by bmwiedemann**
for cct to access its repo from behind a corporate firewall

### CODE REVIEW

**Comment by tpatzig** [comment]
Basically that works. But when setting a proxy, you at least have to exclude from proxy: the cloud domain, the crowbar node admin ip, the endpoint vip in the public net.
It works for me with that:

`printf -v no_proxy '%s,' 192.168.{1..255}.{1..10};`
`no_proxy="localhost,127.0.0.1,.cloud.suse.de,${no_proxy%,}"`

But thats ugly, and every single ip from the range excluded, will be in the no_proxy list.
Better would be, to update the no_proxy variable in mkcloud, if http(s)_proxy is given, to include the admin crowbar ip and the public endpoint vip.

**Comment by bmwiedemann** [comment]
we could address those in a later PR, as this change does not break anything that isnt already broken

File: scripts/qa_crowbarsetup.sh [hunk 1]
```diff
@@ -3133,6 +3134,20 @@ function set_proposalvars
...
+    [[ $http_proxy ]] || [[ $https_proxy ]] || return
+    no_proxy="$adminip,$no_proxy"
+    export no_proxy="${no_proxy%,}";
+    if [[ ! $net_public ]] || [[ $no_proxy =~ ^$net_public ]] ; then
+        return # only apply this once
+    fi
+    local ips
+    printf -v ips '%s,' $net_public.{1..254}
+    no_proxy="$ips$no_proxy"
+    no_proxy="${no_proxy%,}";
```
Comment (tpatzig) [inline scripts/qa_crowbarsetup.sh]: the 2 lines above can be reduced to:
`no_proxy="$no_proxy,${ips%,}";`

File: scripts/qa_crowbarsetup.sh [hunk 2]
```diff
@@ -3133,6 +3134,20 @@ function set_proposalvars
     fi
 }
 
+function set_noproxyvar
+{
+    [[ $http_proxy ]] || [[ $https_proxy ]] || return
+    no_proxy="$adminip,$no_proxy"
+    export no_proxy="${no_proxy%,}";
```
Comment (tpatzig) [inline scripts/qa_crowbarsetup.sh]: should'nt we want to export it at the very end, when $net_public is also included?

File: scripts/qa_crowbarsetup.sh [hunk 3]
```diff
@@ -3133,6 +3134,20 @@ function set_proposalvars
...
+    if [[ ! $net_public ]] || [[ $no_proxy =~ ^$net_public ]] ; then
+        return # only apply this once
```
Comment (tpatzig) [inline scripts/qa_crowbarsetup.sh]: think we also want to check if $adminip is already in

File: scripts/qa_crowbarsetup.sh [hunk 4]
```diff
@@ -5366,6 +5381,7 @@ function onadmin_qa_test
 # $cct_skip_func_tests -> optional, functional tests will be skipped if value is 1, default is 0
 function onadmin_run_cct
 {
+    pre_hook $FUNCNAME
```
Comment (tpatzig) [inline scripts/qa_crowbarsetup.sh]: it is not only needed in the cct step. We need to exclude the admin crowbar IP and the public endpoint vip, right at the beginning. E.g. for tempest it is creating an image and wants to talk to the public vip.

File: scripts/qa_crowbarsetup.sh (hunk 1 above)
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: nay, that can give you a leading comma, if no_proxy is initially empty.

Comment (tpatzig) [inline scripts/qa_crowbarsetup.sh]: ahh yep, convinced 😄

File: scripts/qa_crowbarsetup.sh (hunk 2 above)
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: If I understand bash correctly, `export` just sets a bit and it does not matter where we do that.
Also, net_public might not be set here, so we would not always reach the later part

File: scripts/qa_crowbarsetup.sh (hunk 4 above)
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: that is why we also call it in the very last line in the 'main' part of qa_crowbarsetup

File: scripts/qa_crowbarsetup.sh (hunk 3 above)
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: added a check above

File: scripts/qa_crowbarsetup.sh (hunk 4 above)
Comment (tpatzig) [inline scripts/qa_crowbarsetup.sh]: right 👍

File: scripts/qa_crowbarsetup.sh [hunk 5]
```diff
@@ -3133,6 +3134,21 @@ function set_proposalvars
...
+{
+    [[ $http_proxy ]] || [[ $https_proxy ]] || return 0
+    [[ $admin_ip ]] || return 0
+    [[ $no_proxy =~ "$adminip" ]] || no_proxy="$adminip,$no_proxy"
+    export no_proxy="${no_proxy%,}";
+    if [[ ! $net_public ]] || [[ $no_proxy =~ $net_public ]] ; then
+        return 0 # only apply this once
+    fi
+    local ips
+    printf -v ips '%s,' $net_public.{1..254}
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: I guess this is only for debugging, and should be removed before merging, right?

**Pull request merged**